In [18]:
import numpy as np

### 1. 넘파이를 사용하여 다음과 같은 행렬을 만들고, 이 배열에서 값 70을 인덱싱하세요.  (3.1 - 연습문제 1 변형)
10 20 30 40 <br>
50 60 70 80 <br>

In [3]:
# 밑에 코드 작성
arr1 = np.array([[10,20,30,40],[50,60,70,80]])
print(arr1[1,2], arr1[1][2])

70 70


### 2. 다음 배열은 첫번째 행(row)에 학번, 두번째 행에 영어 성적, 세번째 행에 수학 성적을 적은 배열이다. 영어 성적을 기준으로 각 열(column)을 재정렬하라. (연습문제 3.3.2)

In [28]:
v = np.array([[  1,    2,    3,    4],
       [ 46,   99,  100,   71],
       [ 81,   59,   90,  100]])

# 밑에 코드 작성
eng_rank = np.argsort(v[1])
eng_rank
v[:,[0,3,1,2]]

array([[  1,   4,   2,   3],
       [ 46,  71,  99, 100],
       [ 81, 100,  59,  90]])

### 3. 코사인 유사도 함수 구현하기

In [31]:
def cos_sim(A,B):
    # 밑에 코드 작성
    return np.dot(A,B) / (np.linalg.norm(A) * np.linalg.norm(B))

In [32]:
# Test
doc1 = np.array([0,1,1,1])
doc2 = np.array([1,0,1,1])
result = round(cos_sim(doc1, doc2),2)
result

0.67

- 정답 : 0.67

## 4,5번 문제 : 붓꽃 데이터 활용 

### 4. 붓꽃들의 변수별 가중치를 통해 전체 붓꽃 가중평균 구하기

- 변수별 가중치
    - 꽃받침의 길이 : 0.3, 꽃받침의 폭 : 0.2, 꽃잎의 길이 : 0.25, 꽃잎의 폭 : 0.25

In [38]:
from sklearn.datasets import load_iris

iris = load_iris() # 붓꽃 데이터 로드
X = iris.data

def weighted_average(data):
    
    # 밑에 코드 작성
    weight = [0.3,0.2,0.25,0.25]
    return np.dot(data, weight).mean()

result = weighted_average(X)
result

3.6038

- 정답 <br>
3.6038

### 5. 붓꽃들의 각 변수별 기술통계량 구하기 : 평균, 중앙값, 표준편차

- 소수점 둘째자리까지
- 2차원 리스트의 형태로 반환
    - [[변수 1의 평균, 변수 1의 중앙값, 변수 1의 표준편차], ..., [변수 4의 평균, 변수 4의 중앙값, 변수 4의 표준편차]]

In [60]:
def stats(X):

    # 밑에 코드 작성
    stats = np.column_stack((
        np.mean(X, axis=0), np.median(X, axis=0), np.std(X, axis=0)
    ))

    stats = np.round(stats, 2)
    return stats

result = stats(X)
result



array([[5.84, 5.8 , 0.83],
       [3.06, 3.  , 0.43],
       [3.76, 4.35, 1.76],
       [1.2 , 1.3 , 0.76]])

- 정답 <br>
[[5.84, 5.8, 0.83], [3.06, 3.0, 0.43], [3.76, 4.35, 1.76], [1.2, 1.3, 0.76]]

- numpy 공홈 : https://numpy.org/
- numpy ex100 : https://github.com/rougier/numpy-100/blob/master/100_Numpy_exercises.ipynb